src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs

```diff
@@ -109,8 +108,8 @@ internal void OnIncompleteOperation(int expectedBytesTransferred, int actualByte
         /// <summary>Gets or sets the position within the current stream</summary>
         public sealed override long Position
         {
-            get => _filePosition;
-            set => _filePosition = value;
+            get => Interlocked.Read(ref _filePosition);
+            set => Interlocked.Exchange(ref _filePosition, value);
```

Comment:
What kind of performance impact might this end up having, in particular on 32-bit?  Is this just a case of a test doing something unsupported, or we think this access pattern driving the need for this is valid?

Reply:
>What kind of performance impact might this end up having, in particular on 32-bit? 

For the following micro-benchmark:

```cs
[Benchmark]
[Arguments(OneKibibyte, FileOptions.None)]
[Arguments(OneKibibyte, FileOptions.Asynchronous)]
public void PositionGetSet(long fileSize, FileOptions options)
{
    string filePath = _sourceFilePaths[fileSize];
    using (FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read, FourKibibytes, options))
    {
        for (long i = 0; fileStream.Position < fileSize; i++)
        {
            fileStream.Position = i;
        }
    }
}
```

For x86:
|         Method |        Job | fileSize |      options |       Mean |
|--------------- |----------- |--------- |------------- |-----------:|
| PositionGetSet |     .NET 5 |     1024 |         None |   730.0 us |
| PositionGetSet |     .NET 6 |     1024 |         None |   69.95 us |
| PositionGetSet |    this PR |     1024 |         None |  329.94 us |
|                |            |          |              |            |
| PositionGetSet |     .NET 5 |     1024 | Asynchronous | 2,821.5 us |
| PositionGetSet |     .NET 6 |     1024 | Asynchronous |   80.85 us |
| PositionGetSet |    this PR |     1024 | Asynchronous |  346.32 us |

So the impact on x86 is quite huge, however, we are still few times faster than .NET 5 and I don't expect this method to be on a hot path.

>  Is this just a case of a test doing something unsupported, or we think this access pattern driving the need for this is valid?

The test is not doing anything unsupported. It starts an async write operation and tries to cancel it, if the cancellation succeeds, we check for file length and position to see that they are `0`.

Based on my analysis of the source code and handling of `WriteFile` return values and callbacks my only idea was false sharing and the extra interlocked seems to solve the problem.

I am going to see if marking `_position` as `volatile` would help.

Reply:
> a volatile field cannot be of the type 'long' 

So I don't have other ideas except of disabling this test for x86

Reply:
Actually I have one more idea, but let's wait to see CI results before we get into details (it might be not worth it)